onnxruntime/core/xnnpack/optimizer/xnnpack_transformer.cc

```diff
@@ -0,0 +1,222 @@
+                                        nullptr, "com.microsoft.xnnpack");
+    new_node.AddAttribute("input_padding_top", pads[0]);
+    new_node.AddAttribute("input_padding_right", pads[3]);
+    new_node.AddAttribute("input_padding_bottom", pads[2]);
+    new_node.AddAttribute("input_padding_left", pads[1]);
+
+    new_node.AddAttribute("subsampling_height", strides[0]);
+    new_node.AddAttribute("subsampling_width", strides[1]);
+
+    new_node.AddAttribute("dilation_height", dilations[0]);
+    new_node.AddAttribute("dilation_width", dilations[1]);
+
+    if (!is_depthwise) new_node.AddAttribute("groups", group);
+
+    if (auto_pad_str == "NOTSET") {
+      new_node.AddAttribute("padding_mode", static_cast<int64_t>(0));
+    } else if (auto_pad_str == "VALID") {
+      new_node.AddAttribute("padding_mode", static_cast<int64_t>(0));
+    } else if (auto_pad_str == "SAME_UPPER") {
+      new_node.AddAttribute("padding_mode", static_cast<int64_t>(1));
+    } else {
+      // This line of code should not be reached because in IsConvSupportedByXNNPack function we already checked
+      // auto_pad_str
+      return Status(ONNXRUNTIME, NOT_IMPLEMENTED);
+    }
+
+    ValueMoveInfo value_move_info2(InOutDefSlot{ArgType::kOutput, 0}, ArgType::kOutput, false, false);
+    ORT_RETURN_IF_ERROR(MoveInputOutput(main_graph, *conv_node, new_node, value_move_info2, false));
+    if (!main_graph.RemoveNode(ni)) {
+      return Status(ONNXRUNTIME, FAIL, "remove node failed");
+    }
+    if (bias_node_arg != nullptr) {
+      new_node.MutableInputDefs().push_back(bias_node_arg);
+      new_node.MutableInputArgsCount().push_back(1);
+    }
+
+    if (optimizer_utils::CheckOutputEdges(main_graph, new_node, 1)) {
+      const auto& next_node = *(new_node.OutputNodesBegin());
+      float output_min;
+      float output_max;
+
+      bool has_clip = optimizer_utils::GetClipConstantMinMax(main_graph, next_node, output_min, output_max).IsOK();
+      if (has_clip) {
+        new_node.AddAttribute("output_min", output_min);
+        new_node.AddAttribute("output_max", output_max);
+        ValueMoveInfo value_move_info3(InOutDefSlot{ArgType::kOutput, 0}, InOutDefSlot{ArgType::kOutput, 0});
+        // const_cast
+        Node* clip_node = main_graph.GetNode(next_node.Index());
+        ORT_RETURN_IF_ERROR(MoveInputOutput(main_graph, *clip_node, new_node, value_move_info3, false));
+        if (!main_graph.RemoveNode(next_node.Index())) {
+          return Status(ONNXRUNTIME, FAIL, "remove node failed");
+        }
+      }
+    }
+  }
+
+  if (modified) {
+    ORT_RETURN_IF_ERROR(main_graph.Resolve());
+    auto api_graph = MakeApiGraph(main_graph, cpu_allocator_, kCpuExecutionProvider);
+    // Ignore the return value.
+    onnx_layout_transformation::Optimize(*api_graph, /*allow_extended_ops*/ true);
+  }
```

Comment:
Why do we need to do this here? I wouldn't have expected it was creating a scenario where a Transpose could be optimized.

Reply:
Suppose you have a TF model in NCHW with just one single Conv node. Then you convert it to ONNX. Then the ONNX model will have two Transpose node to transpose input/outputs. Here this xnnpack graph transformer will add new transpose nodes to the graph. Then all the transpose  nodes will be cancelled out after this function call.

The above is just one case.  The onnx_layout_transformation optimizer does more than that.